engine/consensus/matching/core.go

```diff
@@ -946,27 +977,60 @@ func (c *Core) requestPendingReceipts() (int, error) {
 		// get the block header at this height (should not error as heights are finalized)
 		header, err := c.headersDB.ByHeight(height)
 		if err != nil {
-			return 0, fmt.Errorf("could not get header (height=%d): %w", height, err)
+			return 0, 0, fmt.Errorf("could not get header (height=%d): %w", height, err)
 		}
 
 		// check if we have an result for the block at this height
 		blockID := header.ID()
-		if _, ok := knownResultForBlock[blockID]; !ok {
-			missingBlocksOrderedByHeight = append(missingBlocksOrderedByHeight, blockID)
+
+		_, ok := knownResultForBlock[blockID]
+		if ok {
+			continue
 		}
+
+		// since the index is only added when the block which includes the receipts
+		// get finalized, so the returned receipts must be from finalized blocks.
+		// Therefore, the return receipts must be incoporated receipts, which
+		// are safe to be added to the mempool
+		receipts, err := c.receiptsDB.ByBlockIDAllExecutionReceipts(blockID)
+		if err != nil {
+			return 0, 0, fmt.Errorf("could not get receipts by block ID: %v, %w", blockID, err)
+		}
+
+		if len(receipts) > 0 {
+			for _, receipt := range receipts {
+				_, err = c.receipts.AddReceipt(receipt, header)
+				if err != nil {
+					return 0, 0, fmt.Errorf("could not add receipt to receipts mempool %v, %w", receipt.ID(), err)
+				}
+
+				_, err = c.incorporatedResults.Add(
+					flow.NewIncorporatedResult(
+						receipt.ExecutionResult.BlockID,
+						&receipt.ExecutionResult,
+					),
+				)
+
+				if err != nil {
+					return 0, 0, fmt.Errorf("could not add result to incorporated results mempool %v, %w", receipt.ID(), err)
+				}
+			}
+			continue
+		}
```

Comment:
I think in the long term this logic doesn't belong here but in the engine's `OnBlockIncorporated`.
We have talked about this and I understand that this might be a temporary step to consolidate phase 2 of the roadmap.
Hence I would recommend adding this to the comment:
```
// Without the logic below, the matching engine would produce IncorporatedResults
// only from receipts received directly by EN. It would be possible for
// a receipt to come unnoticed by the matching engine if it was
// incorporated in a block by another node and never received directly
// from an EN. Hence we check also if we have the receipts
// in storage (as would have been populated when the block was added to
// the chain).
// NOTE that this logic should eventually be moved in the engine's
// OnBlockIncorporated callback planned for phase 3 of the S&V roadmap,
// and that the IncorporatedResult's IncorporatedBlockID should be set
// correctly.
```

Reply:
thanks Martin. Included your comment